### DISCUSSION TITLE
Lists and tuples not considered array_like -> ravel doesn't work on them?

### DISCUSSION THREAD

**Discussion by nikihowe** [jax-ml/jax#6833] [Q&A] [answered]
In pure numpy, `np.append(np.array([1, 2]), (3, 4))` gives `array([1 2 3 4])`.

In jax numpy, `jnp.append(jnp.array([1, 2]), (3, 4))` gives `TypeError: reshape requires ndarray or scalar arguments, got <class 'tuple'> at position 0.`

This is because `jnp.append` calls `jnp.ravel` on the `(3, 4)`, which checks (and fails the check) whether the tuple is `array_like`. Using a list in place of the tuple also fails the test.

My (possibly incorrect) understanding is that `array_like` in numpy is anything that can be put in a call to `np.array`. It seems that in jax, `array_like` is a property held only by ndarrays, scalars, and objects with the `'__jax_array__'` attribute.

I'm wondering if this is intended behaviour, or if there is a bug here? If the former, then perhaps there should be an explanation in the jax documentation that, unlike in numpy, `array_like` doesn't include tuples or lists.

**Comment by jakevdp** [comment, marked as answer]
Thanks for the question! This is an intended departure from numpy's API: in most places JAX does not silently convert iterables to arrays, because this can hide performance issues.

If you'd like to append a tuple to an array, you can call `jnp.asarray` explicitly; e.g.
```python
tup = (3, 4)
jnp.append(jnp.array([1, 2]), jnp.array(tup))
```
I think we should add a note about this to the [Sharp Bits](https://jax.readthedocs.io/en/latest/notebooks/Common_Gotchas_in_JAX.html) documentation.

**Reply by jakevdp** [reply]
Update (with some discussion) in #6835

Edit: update is now live at https://jax.readthedocs.io/en/latest/notebooks/Common_Gotchas_in_JAX.html#non-array-inputs-numpy-vs-jax

**Reply by nikihowe** [reply]
Thank you!
